src/m365/aad/commands/administrativeunit/administrativeunit-roleassignment-add.spec.ts

```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('fails validation if both role definition id and role definition name are passed', async () => {
+    sinon.stub(cli, 'getSettingWithDefaultValue').callsFake((settingName, defaultValue) => {
+      if (settingName === settingsNames.prompt) {
+        return false;
+      }
+
+      return defaultValue;
+    });
+
+    const actual = await command.validate({
+      options: {
+        administrativeUnitId: administrativeUnitId,
+        roleDefinitionId: roleDefinitionId,
+        roleDefinitionName: roleDefinitionName,
+        userId: userId
+      }
+    }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('fails validation if both administrative unit id and administrative unit name are not passed', async () => {
+    sinon.stub(cli, 'getSettingWithDefaultValue').callsFake((settingName, defaultValue) => {
+      if (settingName === settingsNames.prompt) {
+        return false;
+      }
+
+      return defaultValue;
+    });
+
+    const actual = await command.validate({
+      options: {
+        roleDefinitionId: roleDefinitionId,
+        userId: userId
+      }
+    }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('fails validation if both administrative unit id and administrative unit name are passed', async () => {
+    sinon.stub(cli, 'getSettingWithDefaultValue').callsFake((settingName, defaultValue) => {
+      if (settingName === settingsNames.prompt) {
+        return false;
+      }
+
+      return defaultValue;
+    });
+
+    const actual = await command.validate({
+      options: {
+        administrativeUnitId: administrativeUnitId,
+        administrativeUnitName: administrativeUnitName,
+        roleDefinitionId: roleDefinitionId,
+        userId: userId
+      }
+    }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('fails validation if administrative unit id is not a valid GUID', async () => {
+    const actual = await command.validate({
+      options: {
+        administrativeUnitId: '123',
+        roleDefinitionId: roleDefinitionId,
+        userId: userId
+      }
+    }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('fails validation if role definition id is not a valid GUID', async () => {
+    const actual = await command.validate({
+      options: {
+        administrativeUnitId: administrativeUnitId,
+        roleDefinitionId: '123',
+        userId: userId
+      }
+    }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('fails validation if user id is not a valid GUID', async () => {
+    const actual = await command.validate({
+      options: {
+        administrativeUnitId: administrativeUnitId,
+        roleDefinitionId: roleDefinitionId,
+        userId: '123'
+      }
+    }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('correctly assign a role specified by id to and administrative unit specified by id and to a user specified by id', async () => {
```

Comment:
Seems like we're missing test cases for the different types of failures, such as trying to assign a role to a non existent unit/user/role. Let's add them and make as few assumptions about internal implementation as possible.

Reply:
Added more tests with one generic test which handles OData error.